twill-core/src/main/java/org/apache/twill/kafka/client/KafkaConsumer.java

```diff
@@ -35,12 +35,14 @@
      * Invoked when new messages is available.
      * @param messages Iterator of new messages. The {@link FetchedMessage} instance maybe reused in the Iterator
      *                 and across different invocation.
+     * @param startOffset Offset of the current message to be consumed.
+     * @return The offset of the message to be fetched next.
      */
-    void onReceived(Iterator<FetchedMessage> messages);
+    long onReceived(Iterator<FetchedMessage> messages, long startOffset);
```

Comment:
Can you revert the ordering of the parameters? Logically it fetches from the `startOffset` to produce the `Iterator`.